### ISSUE TITLE
Realm Permission Change callback not Invoked

### ISSUE THREAD

**Issue by zach-snell** [realm/realm-cocoa#4941]
## Goals
A callback to be invoked after the permission change is completed, success or failure so that I can control application logic and flow.

## Expected Results
Callback is invoked regardless of success or failure.

## Actual Results
The callback is not invoked whether a success or a failure occurs.

## Steps to Reproduce
1.  Start a Realm Object Server (**localhost:9080** in the following code)
2. Create a admin user  **u:admin@admin.com**  **p:admin**
3. Run the code contained in TestCase().run()
4. Observe whether or not the callback is invoked.

## Code Sample
```swift
import Foundation
import Realm
import RealmSwift

class TestClass {
    
    let realmObjectServerUrl = URL(string: "http://localhost:9080")
    
    func run() {
        // Credentials for a user that can create a root realm (admin)
        let userCreds = SyncCredentials.usernamePassword(username: "admin@admin.com", password: "admin", register: false)
        
        SyncUser.logIn(with: userCreds, server: realmObjectServerUrl!) { (user, error) in
            
            if(error != nil) {
                print(error!)
                exit(1)
            }
            
            print("Sucessfully logged into realm")
            DispatchQueue.main.async {
                let syncServerURL = URL(string: "realm://localhost:9080/test2")!
                
                let config = Realm.Configuration(syncConfiguration: SyncConfiguration(user: user!, realmURL: syncServerURL), objectTypes: [])
                
                _ = try! Realm(configuration: config)
                
                let writePermissionForTestRealm = RLMSyncPermissionValue(realmPath: "/test2", userID: "*", accessLevel: .write)
                user!.applyPermission(writePermissionForTestRealm) { error in
                    if error != nil {
                        print(error!)
                    }
                    print("permission changed")
                }
            }
        }
    }
}


```

## Version of Realm and Tooling
Realm framework version: ? 
- "realm/realm-cocoa" "v2.7.0"

Realm Object Server version: ?
- Version 1.4.1

Xcode version: ?
- 8.3.2

iOS/OSX version: ?
- 10.12

Dependency manager + version: ?
- Carthage: 0.17.2

**Comment by austinzheng** [comment]
Thanks for getting in touch. I tried reproducing your problem, and was unable to do so. However, I did have to disable App Transport Security in my macOS test application to get things working, since in your sample code you connect to a plain `http` domain. Can you double check that you aren't getting any error messages about App Transport Security when running your application? If there aren't any (you've explicitly disabled it in your Info.plist) and things still don't work, please let me know so I can investigate further.

**Comment by zach-snell** [comment]
@austinzheng The process completes successfully and I can see the changes inside of the realm object browser. However, I am never returned a success or failure message in the callback.

If the protocol was the issue, I feel I would never have completed.

Something that may be helpful to know, I encountered this when testing from a swift console application.

Everything else I have done is working properly though and I assume that to be unrelated.

**Comment by austinzheng** [comment]
I built my sample app on top of Cocoa, so that may account for the difference.

The callback blocks are powered by Realm notifications under the hood, and thus require an active run loop on the thread to work properly. I'm pretty sure Swift console applications don't start a runloop by default; if you start one manually do the callbacks get called?

**Comment by zach-snell** [comment]
@austinzheng, Yes I do, see my updates above. I was editing while you were answering ha.

**Comment by austinzheng** [comment]
Apologies for being dense, what is "yes I do" in response to? Did you start a runloop and still not see the notifications?

**Comment by zach-snell** [comment]
@austinzheng 
That is correct. I start a run loop before the code is ever executed. The code itself is executed within a runloop as shown below.

```swift
DispatchQueue.main.async {
    let test = TestCase()
    test.run()
}

RunLoop.main.run()
```

**Comment by austinzheng** [comment]
Would you be willing to email me your project, so I can poke at it directly? You can send it to `help@realm.io`.

**Comment by zach-snell** [comment]
@austinzheng 

It seems to work now...

In clearing the project to be able to send it over, I ran the test and it did return properly. This is strange as I tested this many times before using the same setup and would not hit breakpoints or print the text specified.

I am not sure why but since it is working, I will not use any more of your time.

**Comment by austinzheng** [comment]
I'd still like to take a look at it, if you're willing to send it over. Regardless, if you run into this issue again please feel free to reopen the ticket or file another one. Thanks for using Realm!

**Comment by zach-snell** [comment]
@austinzheng  I will send you the project if you want since I went through the effort to trim it. However, it works in both the original and the trimmed format now. I am not sure if something was wrong with my RoS or another external issue. 

I have recreated my RoS since then and a few other variables changed.

Thanks though for the timely assistance.

**Comment by zach-snell** [comment]
@austinzheng 
To be clear, would you like the project? And if so would you like it attached here or emailed?

**Comment by austinzheng** [comment]
If it's not too much trouble, I'd like to see it. You can either attach it or email it; whatever is easier.

**Comment by zach-snell** [comment]
[realm_test.zip](https://github.com/realm/realm-cocoa/files/988139/realm_test.zip)

Of course I am not attaching my carthage folder so it will have to be bootstrapped.

**Comment by austinzheng** [comment]
Excellent! Thank you, and best of luck with using Realm.

**Comment by robowen5mac** [comment]
Is this still a potential issue. I am currently using 1.9.2 of the Realm framework

I am seeing similar results in a Mac OS X application. In some cases I can get it to work, but in others, the callback is never called and the I am unable to write to a shared location. I do find that using the deprecated method for setting permissions vis the managementRealm always works.

**Comment by rora002** [comment]
Hi,
I am having the same issue, where I am not seeing the callback run, or the permissions set. Can you please elaborate on what you mean by Realm notifications requiring an active run loop to work properly? Does this mean the functions must be run asynchronously?

Thanks,
Rob

**Comment by oyalhi** [comment]
Having the similar problem. 

1. If the user is the owner of the database
2. and regular user and not admin for the realm
3. and the permission is set by username not by identity

then syncUser.apply(permission) never returns.  If the owner of the database is also an admin OR you set the permission by identity not username, then again, it works as expected.

I hard coded the username (from an existing user) just to test. Below code does NOT work, it never returns form syncUser.apply:
<img width="758" alt="screenshot 2018-03-08 14 02 34" src="https://user-images.githubusercontent.com/11165954/37179254-e92fc220-22d9-11e8-95cf-b5099406830b.png">

Below code does work, only change is setting the permission by identity of the user:
<img width="764" alt="screenshot 2018-03-08 14 03 09" src="https://user-images.githubusercontent.com/11165954/37179358-363817ca-22da-11e8-9a01-1c2f73bdd4de.png">

To sum up, the only way to set permissions is if the user is admin, which does not make sense.
You could try retrieveInfo to get the identifier for the username; however, that also does NOT work unless the syncUser.current is admin.

This is a huge problem.
